src/hierarchies/lapnig_hierarchy.cc

```diff
@@ -0,0 +1,213 @@
+  state.scale = statecast.uni_ls_state().var();
+  set_card(statecast.cardinality());
+}
+
+void LapNIGHierarchy::set_hypers_from_proto(
+    const google::protobuf::Message &hypers_) {
+  auto &hyperscast = downcast_hypers(hypers_).lapnig_state();
+  hypers->mean = hyperscast.mean();
+  hypers->var = hyperscast.var();
+  hypers->scale = hyperscast.scale();
+  hypers->shape = hyperscast.shape();
+  hypers->mh_mean_var = hyperscast.mh_mean_var();
+  hypers->mh_log_scale_var = hyperscast.mh_log_scale_var();
+}
+
+double LapNIGHierarchy::like_lpdf(const Eigen::RowVectorXd &datum) const {
+  return stan::math::double_exponential_lpdf(datum(0), state.mean,
+                                             state.scale);
+}
+
+std::shared_ptr<bayesmix::AlgorithmState::ClusterState>
+LapNIGHierarchy::get_state_proto() const {
+  bayesmix::UniLSState state_;
+  state_.set_mean(state.mean);
+  state_.set_var(state.scale);
+
+  auto out = std::make_shared<bayesmix::AlgorithmState::ClusterState>();
+  out->mutable_uni_ls_state()->CopyFrom(state_);
+  return out;
+}
+
+std::shared_ptr<bayesmix::AlgorithmState::HierarchyHypers>
+LapNIGHierarchy::get_hypers_proto() const {
+  bayesmix::LapNIGState hypers_;
+  hypers_.set_mean(hypers->mean);
+  hypers_.set_var(hypers->var);
+  hypers_.set_shape(hypers->shape);
+  hypers_.set_scale(hypers->scale);
+  hypers_.set_mh_mean_var(hypers->mh_mean_var);
+  hypers_.set_mh_log_scale_var(hypers->mh_log_scale_var);
+
+  auto out = std::make_shared<bayesmix::AlgorithmState::HierarchyHypers>();
+  out->mutable_lapnig_state()->CopyFrom(hypers_);
+  return out;
+}
+
+void LapNIGHierarchy::clear_summary_statistics() {
+  cluster_data_values.clear();
+  sum_abs_diff_curr = 0;
+  sum_abs_diff_prop = 0;
+}
+
+void LapNIGHierarchy::initialize_hypers() {
+  if (prior->has_fixed_values()) {
+    // Set values
+    hypers->mean = prior->fixed_values().mean();
+    hypers->var = prior->fixed_values().var();
+    hypers->shape = prior->fixed_values().shape();
+    hypers->scale = prior->fixed_values().scale();
+    hypers->mh_mean_var = prior->fixed_values().mh_mean_var();
+    hypers->mh_log_scale_var = prior->fixed_values().mh_log_scale_var();
+    // Check validity
+    if (hypers->var <= 0) {
+      throw std::invalid_argument("Variance-scaling parameter must be > 0");
+    }
+    if (hypers->shape <= 0) {
+      throw std::invalid_argument("Shape parameter must be > 0");
+    }
+    if (hypers->scale <= 0) {
+      throw std::invalid_argument("Scale parameter must be > 0");
+    }
+    if (hypers->mh_mean_var <= 0) {
+      throw std::invalid_argument("mh_mean_var parameter must be > 0");
+    }
+    if (hypers->mh_log_scale_var <= 0) {
+      throw std::invalid_argument("mh_log_scale_var parameter must be > 0");
+    }
+  } else {
+    throw std::invalid_argument("Unrecognized hierarchy prior");
+  }
+}
+
+void LapNIGHierarchy::initialize_state() {
+  state.mean = hypers->mean;
+  state.scale = hypers->scale / (hypers->shape + 1);
```

Comment:
This once had a `1.0 /` in front ot it. Is it correct right now? I'm asking just to make sure this is not a typo

Reply:
We corrected it to the mode of an inverse gamma, so it should be correct right now.